File: Solution Recursive.cpp

#include <unordered_set>
#include <vector>
using namespace std;

class Solution {
    
public:
    bool canVisitAllRooms(vector<vector<int>>&rooms) {
        if (rooms.size() <= 1) {
            return true;
        }
        unordered_set<int> visited;
        visited.insert(0);
        return depthFirstSearch_visitRooms(rooms, visited, 0);
    }
    
private:
    bool depthFirstSearch_visitRooms(const vector<vector<int>>&rooms, unordered_set<int>& visited, int currentRoom) {
        const vector<int>& keys = rooms[currentRoom];
        for (const auto& key : keys) {
            if (visited.find(key) == visited.end()) {
                visited.insert(key);
                depthFirstSearch_visitRooms(rooms, visited, key);
            }
        }
        return visited.size() == rooms.size();
    }
};

File: Solution Recursive.java

import java.util.HashSet;
import java.util.List;
import java.util.Set;

public class Solution {

    public boolean canVisitAllRooms(List<List<Integer>> rooms) {
        if (rooms == null || rooms.size() <= 1) {
            return true;
        }
        Set<Integer> visited = new HashSet<>();
        visited.add(0);
        return depthFirstSearch_visitRooms(rooms, visited, 0);
    }

    public boolean depthFirstSearch_visitRooms(List<List<Integer>> rooms, Set<Integer> visited, int currentRoom) {
       final List<Integer> keys = rooms.get(currentRoom);
        for (final int key : keys) {
            if (visited.add(key)) {
                depthFirstSearch_visitRooms(rooms, visited, key);
            }
        }
        return visited.size() == rooms.size();
    }
}
